go-controller/pkg/ovn/default_network_controller.go

```diff
@@ -809,19 +818,37 @@ func (h *defaultNetworkControllerEventHandler) UpdateResource(oldObj, newObj int
 		if !ok {
 			return fmt.Errorf("could not cast oldObj of type %T to *kapi.Node", oldObj)
 		}
-		// determine what actually changed in this update
-		_, nodeSync := h.oc.addNodeFailed.Load(newNode.Name)
-		_, failed := h.oc.nodeClusterRouterPortFailed.Load(newNode.Name)
-		clusterRtrSync := failed || nodeChassisChanged(oldNode, newNode) || nodeSubnetChanged(oldNode, newNode)
-		_, failed = h.oc.mgmtPortFailed.Load(newNode.Name)
-		mgmtSync := failed || macAddressChanged(oldNode, newNode) || nodeSubnetChanged(oldNode, newNode)
-		_, failed = h.oc.gatewaysFailed.Load(newNode.Name)
-		gwSync := (failed || gatewayChanged(oldNode, newNode) ||
-			nodeSubnetChanged(oldNode, newNode) || hostAddressesChanged(oldNode, newNode) ||
-			nodeGatewayMTUSupportChanged(oldNode, newNode))
-		_, hoSync := h.oc.hybridOverlayFailed.Load(newNode.Name)
-
-		return h.oc.addUpdateNodeEvent(newNode, &nodeSyncs{nodeSync, clusterRtrSync, mgmtSync, gwSync, hoSync})
+		if h.oc.isLocalZoneNode(newNode) {
+			var nodeSyncsParam *nodeSyncs
+			if h.oc.isLocalZoneNode(oldNode) {
+				// determine what actually changed in this update
+				_, nodeSync := h.oc.addNodeFailed.Load(newNode.Name)
+				_, failed := h.oc.nodeClusterRouterPortFailed.Load(newNode.Name)
+				clusterRtrSync := failed || nodeChassisChanged(oldNode, newNode) || nodeSubnetChanged(oldNode, newNode)
+				_, failed = h.oc.mgmtPortFailed.Load(newNode.Name)
+				mgmtSync := failed || macAddressChanged(oldNode, newNode) || nodeSubnetChanged(oldNode, newNode)
+				_, failed = h.oc.gatewaysFailed.Load(newNode.Name)
+				gwSync := (failed || gatewayChanged(oldNode, newNode) ||
+					nodeSubnetChanged(oldNode, newNode) || hostAddressesChanged(oldNode, newNode) ||
+					nodeGatewayMTUSupportChanged(oldNode, newNode))
+				_, hoSync := h.oc.hybridOverlayFailed.Load(newNode.Name)
+				nodeSyncsParam = &nodeSyncs{
+					nodeSync,
+					clusterRtrSync,
+					mgmtSync,
+					gwSync,
+					hoSync}
+			} else {
+				klog.Infof("Node %s moved from the remote zone %s to local zone.",
+					newNode.Name, util.GetNodeZone(oldNode), util.GetNodeZone(newNode))
+				// The node is now a local zone node.  Trigger a full node sync.
+				nodeSyncsParam = &nodeSyncs{true, true, true, true, true}
+			}
+
+			return h.oc.addUpdateLocalNodeEvent(newNode, nodeSyncsParam)
+		} else {
```

Comment:
also I think its better to write detailed code comments here on the logic in a tabular fashion which will help both reviewers and anyone who modifies these bits in the future.
```
// +--------------------------+-----------------------+-----------------------+
// |    oldNode               |      newNode          |       Action          |  
// |--------------------------|-----------------------|-----------------------|
// |                          |                       |                       |      
// |         local            |         remote        |          ?            |      
// |                          |                       |                       |       
// |--------------------------|-----------------------|-----------------------|
// |                          |                       |                       |      
// |         local            |         local         |          ?            |    
// |                          |                       |                       |       
// |--------------------------|-----------------------|-----------------------|
// |                          |                       |                       | 
// |         remote           |          local        |          ?            |  
// |                          |                       |                       | 
// |                          |                       |                       | 
// |--------------------------|-----------------------|-----------------------|
// |                          |                       |                       |  
// |       remote             |          remote       |          ?            |  
// |                          |                       |                       |  
// |--------------------------|-----------------------|-----------------------|
```

Reply:
> will take care of it, but since you are keying the cache by nodeName which will be same for oldNode and newNode and now that newNode has moved to local zone this won't delete the oldNode right?

For the scenario you mentioned,  addUpdateRemoteNodeEvent() will be called.

And since the node was local prior to this event, it will be in the localZoneNodes cache and the function oc.deleteNodeEvent(node) will be called.

Reply:
> For the scenario you mentioned, addUpdateRemoteNodeEvent() will be called.
> 
> And since the node was local prior to this event, it will be in the localZoneNodes cache and the function oc.deleteNodeEvent(node) will be called.

ack after reading this comment and https://github.com/ovn-org/ovn-kubernetes/pull/3169#discussion_r1156434475 together it makes sense.